llarp/rpc/lokid_rpc_client.cpp

```diff
@@ -0,0 +1,202 @@
+    }
+
+    void
+    LokidRpcClient::UpdateServiceNodeList()
+    {
+      nlohmann::json request;
+      request["pubkey_ed25519"] = true;
+      request["active_only"] = true;
+      if (not m_CurrentBlockHash.empty())
+        request["poll_block_hash"] = m_CurrentBlockHash;
+      Request(
+          "rpc.get_service_nodes",
+          [self = shared_from_this()](bool success, std::vector<std::string> data) {
+            if (not success)
+            {
+              LogWarn("failed to update service node list");
+              return;
+            }
+            if (data.size() < 2)
+            {
+              LogWarn("lokid gave empty reply for service node list");
+              return;
+            }
+            try
+            {
+              self->HandleGotServiceNodeList(std::move(data[1]));
+            }
+            catch (std::exception& ex)
+            {
+              LogError("failed to process service node list: ", ex.what());
+            }
+          },
+          request.dump());
+    }
+
+    void
+    LokidRpcClient::Connected()
+    {
+      constexpr auto PingInterval = 1min;
+      constexpr auto NodeListUpdateInterval = 30s;
+
+      LogInfo("we connected to lokid [", *m_Connection, "]");
+      Command("admin.lokinet_ping");
+      m_lokiMQ->add_timer(
+          [self = shared_from_this()]() { self->Command("admin.lokinet_ping"); }, PingInterval);
+      m_lokiMQ->add_timer(
+          [self = shared_from_this()]() { self->UpdateServiceNodeList(); }, NodeListUpdateInterval);
+      UpdateServiceNodeList();
+    }
+
+    void
+    LokidRpcClient::HandleGotServiceNodeList(std::string data)
+    {
+      auto j = nlohmann::json::parse(std::move(data));
+      {
+        const auto itr = j.find("block_hash");
+        if (itr != j.end())
+        {
+          m_CurrentBlockHash = itr->get<std::string>();
+        }
+      }
+      {
+        const auto itr = j.find("unchanged");
+        if (itr != j.end())
+        {
+          if (itr->get<bool>())
+          {
+            LogDebug("service node list unchanged");
+            return;
+          }
+        }
+      }
+
+      std::vector<RouterID> nodeList;
+      {
+        const auto itr = j.find("service_node_states");
+        if (itr != j.end() and itr->is_array())
+        {
+          for (auto j_itr = itr->begin(); j_itr != itr->end(); j_itr++)
+          {
+            const auto ed_itr = j_itr->find("pubkey_ed25519");
+            if (ed_itr == j_itr->end() or not ed_itr->is_string())
+              continue;
+            RouterID rid;
+            if (rid.FromHex(ed_itr->get<std::string>()))
+              nodeList.emplace_back(std::move(rid));
+          }
+        }
+      }
+
+      if (nodeList.empty())
+      {
+        LogWarn("got empty service node list from lokid");
+        return;
+      }
+      // inform router about the new list
+      LogicCall(m_Router->logic(), [r = m_Router, nodeList]() { r->SetRouterWhitelist(nodeList); });
```

Comment:
Can move to avoid at least one copy (and maybe two) here:
```C++
[r=m_Router, nodeList=std::move(nodeList)]() mutable { r->SetRouterWhitelist(std::move(nodeList)); });
```